op-chain-ops/cmd/celo-migrate/state.go

```diff
@@ -242,40 +245,53 @@ func applyStateMigrationChanges(config *genesis.DeployConfig, genesis *core.Gene
 // If the code of an existing account is different from the code in the genesis block, it logs a warning.
 // This changes the state root, so `Commit` needs to be called after this function.
-func applyAllocsToState(db *state.StateDB, genesis *core.Genesis, config *params.ChainConfig) {
+func applyAllocsToState(db *state.StateDB, genesis *core.Genesis, config *params.ChainConfig) error {
 	log.Info("Starting to migrate OP contracts into state DB")
 
 	accountCounter := 0
 	overwriteCounter := 0
 	for k, v := range genesis.Alloc {
 		accountCounter++
 
-		balance := uint256.MustFromBig(v.Balance)
+		// Check that the balance of the account to written is zero,
+		// as we must not create new CELo tokens
+		if v.Balance.Cmp(big.NewInt(0)) != 0 {
+			log.Error("Account balance is not zero, would change token supply", "address", k.Hex())
+			return fmt.Errorf("account balance is not zero, would change token supply: %s", k.Hex())
+		}
 
+		nonce := v.Nonce
 		if db.Exist(k) {
-			// If the account already has balance, add it to the balance of the new account
-			balance = balance.Add(balance, db.GetBalance(k))
+			// Check if it's an EOA, if so bail out
+			if db.GetCodeSize(k) == 0 && db.GetNonce(k) > 0 {
+				return fmt.Errorf("account already exists with nonce > 0: %s", k.Hex())
+			}
 
 			currentCode := db.GetCode(k)
 			equalCode := bytes.Equal(currentCode, v.Code)
 			if currentCode != nil && !equalCode {
 				if whitelist, exists := accountOverwriteWhitelist[config.ChainID.Uint64()]; exists {
 					if _, ok := whitelist[k]; ok {
-						log.Info("Account already exists with different code and is whitelisted, overwriting...", "address", k)
+						// keep the existing nonce
+						nonce = db.GetNonce(k)
+						log.Info("Account already exists with different code and is whitelisted, overwriting...", "address", k, "nonce", db.GetNonce(k))
 					} else {
-						log.Warn("Account already exists with different code and is not whitelisted, overwriting...", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
+						log.Error("Account already exists with different code and is not whitelisted, overwriting...", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
+						return fmt.Errorf("account already exists with different code and is not whitelisted: %s", k.Hex())
 					}
 				} else {
-					log.Warn("Account already exists with different code and no whitelist exists", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
+					log.Error("Account already exists with different code and no whitelist exists", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
+					return fmt.Errorf("account already exists with different code and no whitelist exists: %s", k.Hex())
 				}
 
 				overwriteCounter++
 			}
 		}
+
+		// This carries over any existing balance
 		db.CreateAccount(k)
 
-		db.SetNonce(k, v.Nonce)
-		db.SetBalance(k, balance)
+		db.SetNonce(k, nonce)
```

Comment:
This and the SetState calls below, still don't seem correct to me.

So we could have a situation where we have an account that we are deploying that already exists with the same code as the account we are deploying and we will overwrite the nonce and update the associated state. I don't understand why we would want to do this?

To me it seems that if we haven't whitelisted an account to overwrite, we should just fail.

Reply:
Good point.

The logic was hard to follow, so I refactored a bit to make it more clear when accounts are copied. Let me know what you think.